keps/0029-operator-dependencies.md

```diff
@@ -0,0 +1,282 @@
+
+Let's take a look at an example. Here is a simplified operator `AA` with a few dependencies:
+
+```text
+AA
+├── BB
+│   ├── EE
+│   │   ├── H
+│   │   └── I
+│   ├── F
+│   └── GG
+│       ├── J
+│       └── K
+├── CC
+│   ├── L
+│   └── M
+└── D
+
+Legend:
+- Operators and operator-tasks are marked with double letters e.g. 'AA' or `BB`
+- other tasks are marked with single letters e.g. 'D'
+- direct children of an operator are the 'deploy' plan steps e.g. for 'AA' deploy steps are 'BB', 'CC' and 'D'
+```
+
+In the first step we build a dependency graph. A set of all graph vertices (which are task-operators) `S` is defined as `S = {AA, BB, CC, EE, GG}`. A transitive relationship `R` between the vertices is defined as `(a, b) ∈ S` meaning _`a` needs `b` deployed first_. The transitive relationship for the above example is: `R = { (AA,BB), (AA,CC), (BB,EE), (BB,GG) }`. The resulting topological order `O` is therefor `O = (EE, GG, BB, CC, AA)` which has no cycles.
+
+The instance controller (IC) then starts with the execution of the top-level `deploy` plan of the operator `AA`. The first task is the `BB` operator-task. When executing it, IC creates the `Instance-BB` resource and ends current reconciliation. Next, IC notices new `Instance-BB` resource, starts new reconciliation, and executes the  `deploy` plan of the operator `BB` which then creates `Instance-EE` resource. This way we are basically performing the depth-first search for the dependency graph, executing each vertex in the right order e.g. `EE` has to be healthy before `BB` deploy plan can continue with the next step `F`.
+
+We would additionally add the higher-level `Instance` reference (e.g. `AA`) to the `ownerReferences` list of its direct children `Instance`s (e.g. `BB` and `CC`). This would help with determining which `Instance` belongs to which operator and additionally help us with [operator uninstalling](#uninstalling).
+
+The status of the execution can be seen as usual as part of the `Instance.Status`. We could additionally forward the status of a dependency `Instance` to the top-level `Instance.Status` to simplify the overview.
+
+Note that in the above example if e.g. `EE` and `CC` task-operators reference the same operator package they must use distinct instance names `spec.instanceName` so that two separate `Instance`s are deployed. Otherwise, the dependency graph will have a cycle.
+
+##### Dependencies Parametrization
+
+We want to encourage operator composition by providing a way of operator encapsulation. In other words, operator users should not be allowed to arbitrarily modify the parameters of embedded operator instances. The higher-level operator should define all parameters that its **direct** dependency  operators need. Let's demonstrate this on an example of a simple operator `AA` that has operator `BB` as a dependency. 
```

Comment:
Do we plan to restrict the modification of parameters of embedded operators after the initial deployment? i.e. if `BB` was deployed as part of `AA`, can a user use `kudo update --instance BB -p PASSWORD=xyz` or will we detect that `BB` is part of `AA` and prevent that?

Reply:
I don't think there is much point doing that, since the user can always `kudoctl edit instance BB` behind our back?

Reply:
No, if that restriction is in the admission webhook they can't do that.

Reply:
Hm, for the admission webhook to be able to prevent that wouldn't it have to (each time) fetch the original parent instance, and then find and re-execute the relevant `Operator` step to make sure its output matches what the client POSTed? :thinking: 
That strikes me as infeasisble.

Reply:
Yeah, maybe. I don't think it's really a blocker. I think it would be nice to have it in the admission webhook, if it's easily implementable, but if it's too much work or too complicated we can skip it.

Reply:
I think this might be possible to prevent, but I'm not planning on doing that at this point. We've discussed this point with @kensipe and @porridge and there is a lot of value in allowing users to execute plans/update parameters on underlying instances in cases where some valuable features are not exposed at the top level.